src/freelist.rs

```diff
@@ -86,11 +86,12 @@ impl<T: FreeListItem> FreeList<T> {
 
     // TODO(gw): Actually free items from the texture cache!!
     #[allow(dead_code)]
-    pub fn free(&mut self, id: FreeListItemId) {
+    pub fn free(&mut self, id: FreeListItemId) -> &mut T {
```

Comment:
It's not safe to do this - the linked list of free items are stored inside the freed items themselves (for memory / perf reasons), so you can't reference the item after it's freed. Can you change the calling code to get the item first, extract the needed data and then free the item?